tm2py/emme/matrix.py

```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
             Numpy array from OMX file
         """
         return self._omx_file.get_node(path).read()
+
+
+class TollChoiceCalculator:
+    """Implements toll choice calculations.
+
+    Centralized implementation of Toll Choice calculations common to
+    Commercial and Internal-external sub models. Loads input skims
+    from OMXManager
+
+    Args:
+        value_of_time: value of time to use in the utility expression
+        coeff_time: coefficient of time value used in the utility expression
+        operating_cost_per_mile: operating cost value in cents per mile
+            for converting distance to cost
+    """
+
+    def __init__(
+        self, value_of_time: float, coeff_time: float, operating_cost_per_mile: float
+    ):
+        self.value_of_time = value_of_time
+        self.coeff_time = coeff_time
+        self.operating_cost_per_mile = operating_cost_per_mile
+        self._omx_manager = None
+
+    def set_omx_manager(self, omx_manager: OMXManager):
+        """Set the OMX manager for referencing the skim matrices.
+
+        Args:
+            omx_manager: OMXManager which accesses the skim file for reading
+                the skim matrices
+        """
+        self._omx_manager = omx_manager
+
+    def calc_exp_util(
+        self,
+        time_name: str,
+        dist_name: str,
+        toll_names: List[str],
+        toll_factor: float = 1.0,
+    ) -> NumpyArray:
+        """Calculate the exp(utils) for the time, distance and costs skims.
+
+        Loads the referenced skim matrices and calculates the result as:
+        exp(coeff_time * time + coeff_cost * (op_cost * dist + cost)))
+
+        coeff_cost = coeff_time / vot * 0.6
+
+        Args:
+            time_name: Name of the time skim matrix in the OMX file
+            dist_name: Name of the distance skim matrix in the OMX file
+            toll_names: List of names of the the toll skim matrix in the OMX file
+            toll_factor: Optional factor to apply to the tolls
+
+        Returns:
+            A numpy array with the calculated exp(util) result.
+        """
+        vot = self.value_of_time
+        k_ivtt = self.coeff_time
+        op_cost = self.operating_cost_per_mile
+        k_cost = (k_ivtt / vot) * 0.6
+        time = self._omx_manager.read(time_name)
+        dist = self._omx_manager.read(dist_name)
+        toll = self._omx_manager.read(toll_names[0])
+        for name in toll_names[1:]:
+            toll += self._omx_manager.read(name)
+        if toll_factor != 1:
+            toll = toll * toll_factor
+        e_util = exp(k_ivtt * time + k_cost * (op_cost * dist + toll))
+        return e_util
+
+    def mask_non_available(
```

Comment:
This seems like it should be a "general step" for every mode - why is it special for tolls?
2. Shouldn't this be reading in and returning a masked skim?
3. Where are probabilities for tolls zero-ed out? It seems like it should be in same step.